### PR TITLE
Allow custom multi head dispatchers

### PR DESCRIPTION
**Pull request by rciocoiu**

### CODE REVIEW

File: xformers/components/__init__.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
 
 
 def build_multi_head_attention(
-    multi_head_config: Union[MultiHeadDispatchConfig, Dict[str, Any]],
+    multi_head_config: Dict[str, Any],
```
Comment (rciocoiu) [inline xformers/components/__init__.py]: Didn't find any existing use case of calling this with `MultiHeadDispatchConfig` and having this hard dependency here on a specific type of dispatch complicates matters a little bit.

File: xformers/components/__init__.py [hunk 2]
```diff
@@ -38,40 +38,46 @@ def build_multi_head_attention(
...
-                or multi_head_config["attention"]["dim_features"] is None
-            ):
-                multi_head_config["attention"]["dim_features"] = (
-                    multi_head_config["dim_model"] // multi_head_config["num_heads"]
-                )
-
-            multi_head_config["attention"] = build_attention(
-                multi_head_config["attention"]
+    dispatch_class = multi_head_config["dispatch_class"]
+    config_class = multi_head_config["config_class"]
```
Comment (vmarkovtsev) [inline xformers/components/__init__.py]: Let's set backward-compatible defaults here: `MultiHeadDispatch` and `MultiHeadDispatchConfig`.

File: xformers/components/__init__.py (hunk 1 above)
Comment (vmarkovtsev) [reply xformers/components/__init__.py]: We can do this
```
if isinstance(multi_head_config, MultiHeadDispatchConfig):
    multi_head_config = dataclasses.asdict(multi_head_config)
else:
    multi_head_config = multi_head_config.copy()
```

File: xformers/components/__init__.py (hunk 2 above)
Comment (vmarkovtsev) [inline xformers/components/__init__.py]:
```suggestion
    dispatch_class = multi_head_config.pop("dispatch_class", MultiHeadDispatch)
    config_class = multi_head_config.pop("config_class", MultiHeadDispatchConfig)
```

File: xformers/components/__init__.py [hunk 3]
```diff
@@ -38,40 +38,46 @@ def build_multi_head_attention(
...
+        multi_head_config["attention"] = build_attention(
+            multi_head_config["attention"]
+        )
+
+    multi_head_config = config_class(
+        **{
+            k: v for k, v in multi_head_config.items()
+            if k not in ["dispatch_class", "config_class"]
+        },
+    )
```
Comment (vmarkovtsev) [inline xformers/components/__init__.py]:
```suggestion
```
if we do `pop()`

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

**Review by vmarkovtsev — APPROVED** [review]

**Pull request merged**